build/containers/archive-analysis/root/etc/systemd/system/stream_maxlen_config.service

```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Adjust pmproxy.conf stream.maxlen parameter
+Before=pmproxy.service
+
+[Service]
+Type=oneshot
+PassEnvironment=STREAM_MAXLEN_ADJUST
+ExecStart=bash -c 'if [ ! -z $STREAM_MAXLEN_ADJUST ]; then sed -i "s/stream.maxlen = [0-9]*/stream.maxlen = $STREAM_MAXLEN_ADJUST/" /etc/pcp/pmproxy/pmproxy.conf; fi'
```

Comment:
I think this will quickly get out of hand done inline in the service file - let's move this to a pmseries-config.sh script and call that.

Reply:
In fact, considering further - how about we make the tools aware of the PMSERIES_ environment variables directly?  (i.e. override any config file setting with an env var setting).  Then we don't need this additional service configuration systemd magic at all, everything would be simpler I think.

Reply:
Right - we have a well-established convention in several PCP tools where a configuration hierarchy exists.  Command line option overrides are highest priority, then environment variables, then configuration files, then hard-coded default fallback.

My suggestion is to add awareness of PMSERIES_STREAM_MAXLEN and PMSERIES_STREAM_EXPIRE to the libpcp_web library used by pmseries and pmproxy such that these environment variables overrides the default configuration file settings, following this convention.  Since the simplest docker/podman interface seems to be passing in environment variables, this will require no special case code in the Dockerfile - it will Just Work because the tools themselves respect those variables now.

I'm happy to write the C code if it helps.  It seems like a good fit for existing PCP conventions and simplifies the Dockerfile for long-term maintenance IMO.  The downside is users will need to wait for a new PCP release to use it, but I think its best to go for the longer-term cleaner-code solution.

Reply:
+1, I think a generic way to override configuration options from the config file with env vars is a good solution for this issue. Grafana has the same feature: https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#override-configuration-with-environment-variables

In the pcp container I used the `envsubst(1)` utility to override values in the configuration file:
https://github.com/performancecopilot/pcp/blob/e9530dc38827ae31db6bdad0b8012dd1bf4865b9/build/containers/pcp/root/usr/bin/container-entrypoint#L13
https://github.com/performancecopilot/pcp/blob/e9530dc38827ae31db6bdad0b8012dd1bf4865b9/build/containers/pcp/root/usr/share/container-scripts/pcp/pmproxy.conf.template#L74
In this case, the default value moved from the configuration file to the bashscript: https://github.com/performancecopilot/pcp/blob/e9530dc38827ae31db6bdad0b8012dd1bf4865b9/build/containers/pcp/root/usr/bin/container-entrypoint#L7

Reply:
Having the ability to override config variables using PMSERIES_* variables as proposed by @natoscott and @andreasgerstmayr is definitely +1 for me. As Andreas already mentioned, we have this in the Grafana container, where this override is tested by some automated tests. So, having this implemented in the same way as we do it in Grafana will be beneficial for QA as we do not need to develop anything new from the testing point of view.

Reply:
OK, since PCP has a defined hierarchy like that (I didn't know :)), then it makes sense. Since you asked _as generic as possible_, I'd suggest something like this:
~~~
pcp]$ git diff
diff --git a/src/libpcp_web/src/config.c b/src/libpcp_web/src/config.c
index 816bc8813..1dafa0492 100644
--- a/src/libpcp_web/src/config.c
+++ b/src/libpcp_web/src/config.c
@@ -53,12 +53,60 @@ pmIniFileParse(const char *progname, ini_handler handler, void *data)
     return 0;
 }
 
+char *
+construct_canonic_envvar_name(const char *group, const char *key)
+{
+    int i;
+    size_t section_length = strlen(group);
+    size_t option_length = strlen(key);
+    char section[section_length];
+    char option[option_length];
+    char *tmp;
+
+    /* Canonical form: "PCP_<SECTION>_<OPTION>\0" */
+    int envvar_name_len = 4 + section_length + 1 + option_length + 1;
+    char *envvar_name = malloc((size_t)envvar_name_len*sizeof(char));
+    if (!envvar_name)
+       return NULL;
+
+    for (i = 0; i < section_length; i++)
+       section[i] = toupper(group[i]);
+    for (i = 0; i < option_length; i++)
+       if (key[i] == '.')
+               option[i] = '_';
+       else
+               option[i] = toupper(key[i]);
+
+    tmp = mempcpy(envvar_name, "PCP_", 4);
+    tmp = mempcpy(tmp, section, section_length);
+    tmp = mempcpy(tmp, "_", 1);
+    tmp = mempcpy(tmp, option, option_length);
+    envvar_name[envvar_name_len] = '\0';
+
+    return envvar_name;
+}
+
+const char *
+check_envvar_override(const char *group, const char *key, const char *value)
+{
+    char *tmp_value;
+    char *envvar_name = construct_canonic_envvar_name(group, key);
+    if (!envvar_name)
+       goto out_ENOMEM;
+
+    if ((tmp_value = getenv(envvar_name)) != NULL)
+       value = tmp_value;
+
+    free(envvar_name);
+
+out_ENOMEM:
+    return value;
+}
+
 static int
 dict_handler(void *arg, const char *group, const char *key, const char *value)
 {
     dict       *config = (dict *)arg;
     sds                name = sdsempty();
 
+    value = check_envvar_override(group, key, value);
     name = sdscatfmt(name, "%s.%s", group ? group : pmGetProgname(), key);
     if (pmDebugOptions.libweb)
        fprintf(stderr, "pmIniFileParse set %s = %s\n", name, value);
~~~
The "hook" point is in dict_handler(), that is performing the actual add to config dictionary dudring parsing of the config file.. call chain roughly: 
~~~
main@src/pmseries/pmseries.c
-> pmIniFileParse@src/libpcp_web/src/config.c
  -> ini_parse@vendor/github.com/benhoyt/inih/ini.c
    -> ini_parse_file@vendor/github.com/benhoyt/inih/ini.c
      -> ini_parse_stream@vendor/github.com/benhoyt/inih/ini.c
        -> handler == dict_handler@src/libpcp_web/src/config.c
~~~
As part of the handler it checks if an environment variable for the particular config exists and if so, then overrides it from the env var. 
It uses the `group` and `key` to construct the canonical environment variable name of form `PCP_<SECTION>_<OPTION>`, where dots `.` in the option name are represented as underscores `_`.

A sort of unit test of the envvar_name string construction (putting the code into a dedicated `test.c`):
~~~
//    char group[] = "section";
//    char key[] = "option.suboption";

]$ ./test
section len = 7
option len = 16
envvar_name=PCP_SECTION_OPTION_SUBOPTION, len=28
~~~

I know that particular bits, namely the canonic envvar string constuction, ought to be put into some different source file (possibly some header), but I wanted to get your opinion on the approach first. Aesthetics can be done later.

Also, I have yet to try building it from source and testing for PCP actually - will do that tomorrow. Not sure if there are relevant (though still generically standard) libraries already included from some header around, or would be needed to add, namely <ctype.h> for `toupper()` and also `#define _GNU_SOURCE` for `mempcpy()` (the latter can be changed to more generic memcpy, I just went with the more convenient one for now).

Looking forward to your thoughts on the approach. :)

Reply:
| Looking forward to your thoughts on the approach. :)

That's a neat solution - I like it.  I wouldn't lose any sleep over a wasted byte.  :)

mempcpy is a GNU extension, we'll need to adjust that part because PCP is cross-platform - either just use standard interfaces or via autoconf-based ifdef guards and a fallback (the first approach likely simpler).

I also recommend dropping the out_ENOMEM goto, in this case (IMO) the jump adds little value in terms of readability or code simplification - we can just return directly there on failure instead of using goto.

Reply:
So it builds and runs, however it doesn't seem to catch the override (tried direct manual test also). I tried adding some debug prints (even simple printf()), but it seems I misunderstand the callchain and am hooking into the wrong function (my debug prints in `dict_handler@src/libpcp_web/src/config.c` are not even printed ??).

Could you perhaps provide some hint what I'm missing -- see the dev branch for commit https://github.com/vbendel/pcp/tree/envvar_override ? 
Optionally how to work with the `pmDebugOptions.libweb` or other debug options? (I see pmSetConfig() C functions - are those the only interface to enable the debug points?)

Reply:
@vbendel yes, libpcp_web is a shared library - your code does appear to be invoked:

```
$ export LD_LIBRARY_PATH=`pwd`/../libpcp_web/src
$ ./pmseries hinv.ncpu
DEBUG: group=pmproxy key=pcp.enabled value=true
DEBUG: group=pmproxy key=http.enabled value=true
DEBUG: group=pmproxy key=redis.enabled value=true
DEBUG: group=pmproxy key=secure.enabled value=true
DEBUG: group=redis key=enabled value=true
DEBUG: group=redis key=servers value=localhost:6379
DEBUG: group=discover key=enabled value=true
DEBUG: group=discover key=exclude.metrics value=proc.*,acct.*
DEBUG: group=discover key=exclude.indoms value=3.9,3.40,79.7
DEBUG: group=pmsearch key=enabled value=true
DEBUG: group=pmsearch key=count value=10
DEBUG: group=pmseries key=enabled value=true
DEBUG: group=pmseries key=cursor.count value=256
DEBUG: group=pmseries key=stream.expire value=86400
DEBUG: group=pmseries key=stream.maxlen value=8640
6b01fd0761645cbccb6312e0cf3bb9c838dfe7af
a55d8e2a7a7ef7e0f70e8ab6b1ce767cd3a0ce0b
cbf17c2f8421d6a1c96b64dc582a58ef177bccbe
```

You can access the debugging flags through the -D/--debug option available with all PCP binaries, e.g.

```
$ ./pmseries -Dlibweb hinv.ncpu
DEBUG: group=pmproxy key=pcp.enabled value=true
pmIniFileParse set pmproxy.pcp.enabled = true
DEBUG: group=pmproxy key=http.enabled value=true
pmIniFileParse set pmproxy.http.enabled = true
DEBUG: group=pmproxy key=redis.enabled value=true
pmIniFileParse set pmproxy.redis.enabled = true
DEBUG: group=pmproxy key=secure.enabled value=true
pmIniFileParse set pmproxy.secure.enabled = true
DEBUG: group=redis key=enabled value=true
pmIniFileParse set redis.enabled = true
DEBUG: group=redis key=servers value=localhost:6379
pmIniFileParse set redis.servers = localhost:6379
DEBUG: group=discover key=enabled value=true
pmIniFileParse set discover.enabled = true
DEBUG: group=discover key=exclude.metrics value=proc.*,acct.*
pmIniFileParse set discover.exclude.metrics = proc.*,acct.*
DEBUG: group=discover key=exclude.indoms value=3.9,3.40,79.7
pmIniFileParse set discover.exclude.indoms = 3.9,3.40,79.7
DEBUG: group=pmsearch key=enabled value=true
pmIniFileParse set pmsearch.enabled = true
DEBUG: group=pmsearch key=count value=10
pmIniFileParse set pmsearch.count = 10
DEBUG: group=pmseries key=enabled value=true
pmIniFileParse set pmseries.enabled = true
DEBUG: group=pmseries key=cursor.count value=256
pmIniFileParse set pmseries.cursor.count = 256
DEBUG: group=pmseries key=stream.expire value=86400
pmIniFileParse set pmseries.stream.expire = 86400
DEBUG: group=pmseries key=stream.maxlen value=8640
pmIniFileParse set pmseries.stream.maxlen = 8640
6b01fd0761645cbccb6312e0cf3bb9c838dfe7af
a55d8e2a7a7ef7e0f70e8ab6b1ce767cd3a0ce0b
cbf17c2f8421d6a1c96b64dc582a58ef177bccbe
```

pmdbg -l/--list for the full list - multiple debug settings can be used too via comma-separated list.

Reply:
Thank you for the hints!

I seems to be working:
~~~
[root@localhost pcp]# export PCP_PMSEARCH_COUNT=12
[root@localhost pcp]# export PCP_PMSERIES_STREAM_MAXLEN=456789
[root@localhost pcp]# ./src/pmseries/pmseries -Dlibweb --load <archive> 
pmIniFileParse set pmproxy.pcp.enabled = true
pmIniFileParse set pmproxy.http.enabled = true
pmIniFileParse set pmproxy.redis.enabled = true
pmIniFileParse set pmproxy.secure.enabled = true
pmIniFileParse set redis.enabled = true
pmIniFileParse set redis.servers = localhost:6379
pmIniFileParse set discover.enabled = true
pmIniFileParse set discover.exclude.metrics = proc.*,acct.*
pmIniFileParse set discover.exclude.indoms = 3.9,3.40,79.7
pmIniFileParse set pmsearch.enabled = true
pmIniFileParse set pmsearch.count = 12   <<<---
pmIniFileParse set pmseries.enabled = true
pmIniFileParse set pmseries.cursor.count = 256
pmIniFileParse set pmseries.stream.expire = 86400
pmIniFileParse set pmseries.stream.maxlen = 456789  <<<---
~~~

Also tested in practice and can confirm the records are not overwritten in Redis DB, i.e. the updated XADD MAXLEN param is indeed properly used.

Now the question is where to put the `construct_canonic_envvar_name()` function (needs also `<ctype.h>`)? Should we:
- Create a `src/libpcp_web/src/config.h` (only config.c is there atm); or
- Put it somewhere else, like general `libpcp.h` for example (or other suggestions); or
- Leave it simply as a "local" function in `src/libpcp_web/src/config.c`, as depicted right now?
(Personally I'd create a new `src/libpcp_web/src/config.h` for it.)

Looking forward to your suggestions. Afterwards I'll take your feedback and update this PR overall.